common.py

```diff
@@ -0,0 +1,235 @@
+    np.all(bbox[:, 1] < -1) or np.all(bbox[:, 1] > 1) and
+    np.all(bbox[:, 2] < -1) or np.all(bbox[:, 2] > 1)):
+    if not info.obj.use_f3d_culling: # if obj is not meant to be culled in game, apply all materials
+      for mat_idx, indices_count, f64mat in info.mats: render_state.set_if_not_none(f64mat.state)
+    return
+
+  render_engine.shader.uniform_float("matMVP", info.mvp_matrix)
+  render_engine.shader.uniform_float("matNorm", info.normal_matrix)
+
+  for mat_idx, indices_count, f64mat in info.mats:
+    render_state.set_if_not_none(f64mat.state)
+
+    gpu.state.face_culling_set(f64mat.cull)
+    if not render_engine.use_atomic_rendering:
+      gpu.state.blend_set(render_state.render_mode.blend)
+      gpu.state.depth_test_set(render_state.render_mode.depth_test)
+      gpu.state.depth_mask_set(render_state.render_mode.depth_write)
+
+    for i in range(8):
+      if render_state.tex_confs[i].buff is not render_engine.last_used_textures.get(i): 
+        render_engine.shader.uniform_sampler(f"tex{i}", render_state.tex_confs[i].buff)
+        render_engine.last_used_textures[i] = render_state.tex_confs[i].buff
+
+    light_data = []
+    for l in render_state.lights[:render_state.light_count]:
+      light_data.extend(l.color)
+      light_data.extend(l.direction)
+    light_data.extend([0.0] * ((8 - render_state.light_count) * 7))
+
+    tex_data = []
+    for t in render_state.tex_confs: tex_data.extend(t.values)
+
+    info.render_obj.mat_data[mat_idx] = UNIFORM_BUFFER_STRUCT.pack(
+      *tex_data,
+      *light_data,
+      *render_state.render_mode.blender,
+      *render_state.cc,
+      f64mat.geo_mode,
+      f64mat.othermode_l,
+      f64mat.othermode_h,
+      f64mat.flags | render_state.render_mode.flags,
+      *render_state.prim_color,
+      *render_state.prim_lod,
+      *f64mat.prim_depth,
+      *render_state.env_color,
+      *render_state.ambient_color,
+      *render_state.ck[:3],
+      render_state.alpha_clip,
+      *render_state.ck[3:6],
+      render_state.light_count,
+      *render_state.ck[6:9],
+      f64mat.uv_basis,
+      *render_state.convert,
+      f64mat.mip_count,
+    )
+    
+    info.render_obj.ubo_mat_data[mat_idx].update(info.render_obj.mat_data[mat_idx])                        
+    render_engine.shader.uniform_block("material", info.render_obj.ubo_mat_data[mat_idx])
+
+    if render_engine.draw_range_impl:
+      info.render_obj.batch.draw_range(render_engine.shader, elem_start=info.render_obj.index_offsets[mat_idx] * 3, elem_count=indices_count)
+    else:
+      info.render_obj.batch[mat_idx].draw(render_engine.shader)
+
+def collect_obj_info(render_engine: "Fast64RenderEngine", obj: bpy.types.Object, depsgraph: bpy.types.Depsgraph, hidden_objs: set[bpy.types.Object], space_view_3d: bpy.types.SpaceView3D, projection_matrix: mathutils.Matrix, view_matrix: mathutils.Matrix, always_set: bool, set_light_dir = True):
+  global F64_GLOBALS
```

Comment:
there's no need to keep using `global F64_GLOBALS`, you never assign to that variable directly but only to members inside it